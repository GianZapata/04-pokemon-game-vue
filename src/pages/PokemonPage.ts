import { defineComponent } from "vue"

import PokemonImage from "@/components/PokemonImage.vue"
import PokemonOptions from "@/components/PokemonOptions.vue"
import { usePokemons } from "@/composables/usePokemons"

export default defineComponent({
  name: "PokemonPage",
  components: {
    PokemonImage,
    PokemonOptions,
  },
  setup() {
    const {
      /** Properties */
      message,
      pokemon,
      pokemons,
      showAnswer,
      showPokemon,

      /** Methods */
      checkAnswer,
      mixPokemons,
      newGame,
    } = usePokemons()

    mixPokemons()

    return {
      message,
      pokemon,
      pokemons,
      showAnswer,
      showPokemon,

      /** Methods */
      checkAnswer,
      mixPokemons,
      newGame,
    }
  },
})
