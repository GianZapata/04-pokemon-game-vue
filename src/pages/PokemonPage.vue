<template>
  <div class="flex flex-col items-center justify-center" v-if="!pokemon">
    <h1 class="text-3xl text-center mt-20 font-black text-slate-900">
      Cargando ...
    </h1>
  </div>

  <div class="flex flex-col items-center justify-center" v-else>
    <h1 class="text-3xl text-center mt-20 font-black text-slate-900">
      ¿Quién este este pokemon?
    </h1>
    <PokemonImage />
    <PokemonOptions />

    <template v-if="showAnswer">
      <h2
        class="text-2xl text-center mt-20 font-black text-slate-900"
        v-if="message"
      >
        {{ message }}
      </h2>

      <button
        class="font-bold py-2 px-4 rounded w-96 my-2 transition duration-500 ease-in-out transform hover:-translate-y-1 hover:scale-110 bg-blue-500 hover:bg-blue-700 text-white"
        @click="newGame"
      >
        Nuevo juego
      </button>
    </template>
  </div>
</template>

<script setup lang="ts">
import PokemonOptions from "@/components/PokemonOptions.vue"
import PokemonImage from "@/components/PokemonImage.vue"

import { usePokemons } from "@/composables/usePokemons"

const {
  /** Properties */
  message,
  pokemon,
  showAnswer,

  /** Methods */
  mixPokemons,
  newGame,
} = usePokemons()
mixPokemons()
</script>
