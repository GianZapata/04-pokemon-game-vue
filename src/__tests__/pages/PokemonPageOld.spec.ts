import type { ComponentPublicInstance } from "vue"
import { shallowMount, VueWrapper } from "@vue/test-utils"
import { beforeEach, describe, test, expect, vi } from "vitest"
import PokemonPage from "@/pages/PokemonPageOld.vue"
import { mockedPokemons } from "../mocks/pokemons.mock"

describe("PokemonPage Component", () => {
  let wrapper: VueWrapper<ComponentPublicInstance>

  beforeEach(() => {
    wrapper = shallowMount(PokemonPage)
  })

  test("Debe de hacer el snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  test("Debe de llamar mixPokemons al montar", () => {
    const mixPokemonsSpy = vi.spyOn<any, any>(
      PokemonPage.methods,
      PokemonPage.methods?.mixPokemons.name
    )
    shallowMount(PokemonPage)
    expect(mixPokemonsSpy).toHaveBeenCalled()
  })

  test("Debe de hacer match con el snapshot cuando cargan los pokemons", () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemons: mockedPokemons,
          pokemon: mockedPokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: "",
        }
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  test("Debe de mostrar los componentes de PokemonPicture y PokemonOptions", () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemons: mockedPokemons,
          pokemon: mockedPokemons[0],
          showPokemon: true,
          showAnswer: true,
          message: "",
        }
      },
    })

    const image = wrapper.find("pokemon-image-stub")
    const options = wrapper.find("pokemon-options-stub")

    expect(image.exists()).toBeTruthy()
    expect(options.exists()).toBeTruthy()
    expect(image.attributes("pokemonid")).toBe("1")
    expect(options.attributes("pokemons")).toBeTruthy()
  })

  test("Pruebas con checkAnswer", async () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemons: mockedPokemons,
          pokemon: mockedPokemons[0],
          showPokemon: true,
          showAnswer: true,
          message: "",
        }
      },
    })

    await wrapper.vm.checkAnswer(1)

    expect(wrapper.find("h2").exists()).toBeTruthy()
    expect(wrapper.vm.showPokemon).toBeTruthy()
    expect(wrapper.find("h2").text()).toBe(
      `Â¡Correcto! El pokemon es ${mockedPokemons[0].name}`
    )

    await wrapper.vm.checkAnswer(2)

    expect(wrapper.vm.message).toBe(
      `Oops! El pokemon era ${mockedPokemons[0].name}`
    )
  })
})
