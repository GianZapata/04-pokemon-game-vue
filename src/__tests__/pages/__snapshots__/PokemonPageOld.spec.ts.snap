// Vitest Snapshot v1

exports[`PokemonPage Component > Debe de hacer el snapshot 1`] = `
"<div class=\\"flex flex-col items-center justify-center\\">
  <h1 class=\\"text-3xl text-center mt-20 font-black text-slate-900\\"> Cargando ... </h1>
</div>"
`;

exports[`PokemonPage Component > Debe de hacer match con el snapshot cuando cargan los pokemons 1`] = `
"<div class=\\"flex flex-col items-center justify-center\\">
  <h1 class=\\"text-3xl text-center mt-20 font-black text-slate-900\\"> ¿Quién este este pokemon? </h1>
  <pokemon-image-stub pokemonid=\\"1\\" showpokemon=\\"false\\"></pokemon-image-stub>
  <pokemon-options-stub pokemons=\\"[object Object],[object Object],[object Object],[object Object],[object Object]\\"></pokemon-options-stub>
  <!--v-if-->
</div>"
`;
